### PR TITLE
DBZ-6858 Timestamp metadata for watermarking signals

### PR DESCRIPTION
**Pull request by ani-sha**
https://issues.redhat.com/browse/DBZ-6858

### CODE REVIEW

**Comment by jpechane** [comment]
Please as was proposed, make this solution more generic with a JSON formatted metadata approach. Maybe it would be possible to extend `IncrementalSnapshotContext` or introduce a new one. The class will provide a method `signalMetadataString()` that will return a formatted JSON which will be used in database write. Currently two fields will be supported

* `openWindowTimestamp` - available at both open and close window signals. For close window signal it will be timestamp correlated with associated open window signal
* `closeWindowTimestamp` - available only at the close window signal

WDYT?

**Comment by ani-sha** [comment]
@jpechane Yes, I am working on this. Can you please clarify when and where should the timestamp be initialised/declared? Is it `emitWindowOpen()`'s `prepareUpdate` statement for the `openWindowTimestamp` or prior to that? I had this confusion just to ensure the accurate timestamp is being written into the table.

**Comment by jpechane** [comment]
@ani-sha In general the timestamp should be created as near to the point of write as possible. But TBH I don't think it is really a big deal here. The value does not need to be exact, the trend is most important so as the values will be created always at the same place they'll be consistent.

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/SignalMetadata.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    private final String openWindowTimestamp;
+    private final String closeWindowTimestamp;
+
+    public SignalMetadata(String openWindowTimestamp, String closeWindowTimestamp) {
+        this.openWindowTimestamp = openWindowTimestamp;
+        this.closeWindowTimestamp = closeWindowTimestamp;
+    }
+
+    public String signalMetadataString(SignalType type) {
```
[the remark this answers is not in the document]
Comment (ani-sha) [reply debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/SignalMetadata.java:22]: Yes, that should be fine.

File: debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/SignalMetadata.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.pipeline.source.snapshot.incremental;
+
+public class SignalMetadata {
```
Comment (jpechane) [inline debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/SignalMetadata.java:8]: Please add JavaDoc.

**Review by jpechane — COMMENTED** [review]
@ani-sha I am fine with the solution, thanks! I left couple of comments that relates to the small logical code changes.

@mfvitale It is about time for us to introduce some unit test level testing of incremental snapshot code for functionlity like this. Do you have any idea how we could introduce it easily?

**Comment by mfvitale** [comment]
> @mfvitale It is about time for us to introduce some unit test level testing of incremental snapshot code for functionlity like this. Do you have any idea how we could introduce it easily?

I think it depends case by case. In this one, an easy option is to unit test the `SignalMetadata` and then try with `SignalBasedIncrementalSnapshotChangeEventSource` and try to mock the connection and then `verify` on the mock the correct method calls, maybe in that specific case `ArgumentCaptor` must be used.

**Comment by jpechane** [comment]
@ani-sha Are you sure you pushed the changes? I see the comments being resolved but the code stays the same.

**Comment by ani-sha** [comment]
Sorry @jpechane, I had pushed the changes in the wrong branch. Added them now!

**Comment by jpechane** [comment]
@ani-sha Applied, thanks! As we spoke on the call, please create a follow-up task for doing a test for this one that could be expanded later for other unit tests.

**Pull request merged**